python-package/setup.py

```diff
@@ -276,7 +276,7 @@ def run(self):
           install_requires=[
               'numpy',
               'scipy',
-              'scikit-learn<=0.21.3'
+              'scikit-learn!=0.22.0'
```

Comment:
is this `'scikit-learn~=0.22,!=0.22.0'`?

Reply:
Sorry, I can't find `~` in the `setuptools` docs. Can you please clarify what do you mean?
> A version specifier is one of the operators <, >, <=, >=, == or !=, followed by a version identifier.
    https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-dependencies

Reply:
Version specifiers
https://www.python.org/dev/peps/pep-0440/#version-specifiers

Reply:
OK. I guess that this specifier is not compatible with `setuptools` as it is not listed there, is it?

Also, why do we need it? We are compatible with all (reasonable) versions except `0.22.0`.

Reply:
ping @henry0312

Reply:
> OK. I guess that this specifier is not compatible with setuptools as it is not listed there, is it?

I don't know the specific version but `setuptools` used in sklearn, numpy, scipy and pandas supports version specifier as least.

> Also, why do we need it? We are compatible with all (reasonable) versions except 0.22.0.

Not enough.
As you know, many python packages follow [Semantic Versioning](https://semver.org/), including scikit-learn.

If we specify `scikit-learn!=0.22.0` and v1.0.0 of scikit-learn is released, LightGBM will allow users to use v1.0.0, which can lead to break backward compatibility.
Therefore, if we follow [Semantic Versioning](https://semver.org/), we should specify libraries' version by Version specifiers.

Reply:
> If we specify scikit-learn!=0.22.0 and v1.0.0 of scikit-learn is released, LightGBM will allow users to use v1.0.0,

I don't see any problems with it. Moreover, I believe that this should be an expected behavior.
Before now we have never blocked **future** versions of libraries, assuming that by default we are compatible with them. If it is not true, someone comes and posts an issue and we are investigating.
In case of prohibiting future releases it is highly likely that we'll forget to update those requirements or will do it with huge delay making user experience inconvenient.

Reply:
> Before now we have never blocked future versions of libraries, assuming that by default we are compatible with them. 

When that is our policy, as you say, there is no problem and I'll approve this PR, though I don't think the policy is common in software engineering.
In other words, a new version which can breaks backward compatibility in terms of Semantic Versioning is released, we should check its compatibility and will add suuport of it in series.

Reply:
So I'll approve this when our support policy is clear.

Reply:
In fact, if we specified Sphinx version, https://github.com/microsoft/LightGBM/commit/ce08242567e1e37e1f947038a78ac671c98ed18e#diff-e0c7b285f2f2d8028247b8cfd4dc76fcR1 were not needed, though that are just tests.

I think unless we follow the basic rules of software engineering, the quality of library won't increase so much even If we add more and more tests.

So I'll approve this when our support policy is clear.

Reply:
@henry0312 Actually it's pretty common in software engineering to allow for future versions of dependencies in **libraries** (while it's common to be more strict in end-user products). If a dependency follows Semantic Versioning (SemVer), you could put a restriction on major versions (e.g. scikit-learn<1.0), but I don't think sklearn is following SemVer. Unless you're ready to release a new version of LightGBM every time a new minor version or patch for a dependency is released, of course.

Reply:
@henry0312 
> though I don't think the policy is common in software engineering.

It's kind of debatable statements. And I don't think that this PR is a good place to discuss it. I think we should create a separate issue where we should at least get opinions of other maintainers before changing the current policy.

> we should check its compatibility and will add suuport of it in series.

In ideal situation, yes! You are absolutely right! But given that we are lacking human resources to do it in time, I suppose that it will simply abandoned in the near future. Which will result in very old dependencies of LightGBM.

Reply:
> Unless you're ready to release a new version of LightGBM every time a new minor version or patch for a dependency is released, of course.

Absolutely agree! It is definitely not our case when only 2-3 maintainers work actively at the same time.

Reply:
@ekerazha @StrikerRUS I don't mean all of the future versions wont' be allowed, but only the future versions that can breaks backward compatibility (in other words, a release changes major version), which version specifiers (i.e. `scikit-learn~=0.22,!=0.22.0` or `scikit-learn>=0.22, == 0.*`) realizes.
And is LightGBM an almost end-user products? I think so.

> I don't think sklearn is following SemVer.

oh, really...? Actually, I'm not sure.

Reply:
@henry0312 
> And is LightGBM an almost end-user products? I think so.
LightGBM is a library, a travel website which uses LightGBM to recommend travels would be the end-user product.

> oh, really...? Actually, I'm not sure.
scikit-learn had breaking API changes in minor versions, so it's not following SemVer.

Reply:
@henry0312 I guess `0.22` is a **minor** release, but it broke  backward compatibility.

Again, please let's move all our discussion to a separate issue where everyone will be able to comment. It's very hard to find it here and I want to merge this PR ASAP to enable the latest scikit-learn support.